### PR TITLE
Kotlin Conversion: ServiceFactory

### PR DESCRIPTION
**Pull request by sharvaniharan**

### CODE REVIEW

File: app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            .client(okHttpClientBuilder.build())
+            .baseUrl(baseUrl)
+            .client(client.newBuilder().addInterceptor(LanguageVariantHeaderInterceptor(wiki)).build())
+            .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
+            .addConverterFactory(GsonConverterFactory.create(GsonUtil.getDefaultGson()))
+            .build()
+    }
+
+    private class LanguageVariantHeaderInterceptor(private val wiki: WikiSite?) :
+        Interceptor {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt:107]: Could these be in one line?

File: app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    @JvmStatic
+    fun getAnalyticsRest(streamConfig: StreamConfig): EventService {
+        val destinationEventService = streamConfig.destinationEventService
+        if (ANALYTICS_REST_SERVICE_CACHE[destinationEventService.id] != null) {
+            return ANALYTICS_REST_SERVICE_CACHE[destinationEventService.id]!!
+        }
+        val intakeBaseUriOverride = Prefs.getEventPlatformIntakeUriOverride()
+        val r = createRetrofit(null, StringUtils.defaultString(intakeBaseUriOverride, destinationEventService.baseUri))
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt:68]: Considering using `orEmpty().ifEmpty {} `? e.g.: `Prefs.getEventPlatformIntakeUriOverride().orEmpty().ifEmpty { destinationEventService.baseUri }`

**Review by cooltey — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            path += "/"
+        }
+        return path
+    }
+
+    private fun createRetrofit(wiki: WikiSite?, baseUrl: String): Retrofit {
+        val okHttpClientBuilder = client.newBuilder()
+        if (wiki != null) {
+            okHttpClientBuilder.addInterceptor(LanguageVariantHeaderInterceptor(wiki))
+        }
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt:95]: Very curious that why we have to add the `LanguageVariantHeaderInterceptor` here, and then add it again below? Do you have any reason why we are doing this?

File: app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        return Retrofit.Builder()
+            .client(okHttpClientBuilder.build())
+            .baseUrl(baseUrl)
+            .client(client.newBuilder().addInterceptor(LanguageVariantHeaderInterceptor(wiki)).build())
+            .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
+            .addConverterFactory(GsonConverterFactory.create(GsonUtil.getDefaultGson()))
+            .build()
+    }
+
+    private class LanguageVariantHeaderInterceptor(private val wiki: WikiSite?) : Interceptor {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt:105]: Looks like if the `wiki` is null, then the entire process is unreachable. (or maybe will lead to a crash). Should we move the null check logic into this method?

**Review by cooltey — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt (hunk 4 above)
Comment (sharvaniharan) [reply app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt:102]: Good catch @cooltey ! done!

**Pull request merged**